packages/host/cpp/ThreadsafeFunction.cpp

```diff
@@ -0,0 +1,278 @@
+      napi_fatal_error("ThreadSafeFunction::ThreadSafeFunction",
+          NAPI_AUTO_LENGTH,
+          "Failed to create JS function reference",
+          NAPI_AUTO_LENGTH);
+    }
+  }
+}
+
+ThreadSafeFunction::~ThreadSafeFunction() {
+  if (jsFuncRef_) {
+    napi_delete_reference(env_, jsFuncRef_);
+  }
+}
+
+std::shared_ptr<ThreadSafeFunction> ThreadSafeFunction::create(
+    std::weak_ptr<facebook::react::CallInvoker> callInvoker,
+    napi_env env,
+    napi_value jsFunc,
+    napi_value asyncResource,
+    napi_value asyncResourceName,
+    size_t maxQueueSize,
+    size_t initialThreadCount,
+    void* threadFinalizeData,
+    napi_finalize threadFinalizeCb,
+    void* context,
+    napi_threadsafe_function_call_js callJsCb) {
+  const auto function =
+      std::make_shared<ThreadSafeFunction>(std::move(callInvoker),
+          env,
+          jsFunc,
+          asyncResource,
+          asyncResourceName,
+          maxQueueSize,
+          initialThreadCount,
+          threadFinalizeData,
+          threadFinalizeCb,
+          context,
+          callJsCb);
+
+  {
+    std::lock_guard lock{registryMutex};
+    registry[function->id_] = function;
+  }
+
+  return function;
+}
+
+std::shared_ptr<ThreadSafeFunction> ThreadSafeFunction::get(
+    napi_threadsafe_function func) {
+  std::lock_guard lock{registryMutex};
+  // Cast the handle back to ID for registry lookup
+  const auto id = reinterpret_cast<std::uintptr_t>(func);
+  const auto it = registry.find(id);
+  return it != registry.end() ? it->second : nullptr;
+}
+
+napi_threadsafe_function ThreadSafeFunction::getHandle() const noexcept {
+  return reinterpret_cast<napi_threadsafe_function>(id_);
+}
+
+napi_status ThreadSafeFunction::getContext(void** result) noexcept {
+  if (!result) {
+    return napi_invalid_arg;
+  }
+
+  *result = context_;
+  return napi_ok;
+}
+
+napi_status ThreadSafeFunction::call(
+    void* data, napi_threadsafe_function_call_mode isBlocking) noexcept {
+  if (isClosingOrAborted()) {
+    return napi_closing;
+  }
+
+  {
+    std::unique_lock lock{queueMutex_};
+    // Backpressure: enforce maxQueueSize_. If nonblocking, fail fast; if
+    // blocking, wait until space is available or closing/aborted.
+    if (maxQueueSize_ && queue_.size() >= maxQueueSize_) {
+      if (isBlocking == napi_tsfn_nonblocking) {
+        return napi_queue_full;
+      }
+      queueCv_.wait(lock, [&] {
+        return queue_.size() < maxQueueSize_ || isClosingOrAborted();
+      });
+      if (isClosingOrAborted()) return napi_closing;
+    }
+    queue_.push(data);
+  }
+
+  const auto invoker = callInvoker_.lock();
+  if (!invoker) {
+    log_debug("Error: No CallInvoker available for ThreadSafeFunction");
+    return napi_generic_failure;
+  }
+  // Hop to JS thread; we drain one item per hop to keep latency predictable
+  // and avoid long monopolization of the JS queue.
+  invoker->invokeAsync([self = shared_from_this()] { self->processQueue(); });
```

Comment:
This one-line change was enough to get thread-hopping in NativeScript Node-API working (where before, it was crashing):

```suggestion
  invoker->invokeSync([self = shared_from_this()] { self->processQueue(); });
```

I was able to do intricate moves back and forth between the JS and main threads, and use JS timers (see screenshot below). It all worked perfectly.

<img width="481" height="995" alt="image" src="https://github.com/user-attachments/assets/c3e994b1-9c63-4855-be77-af1000d8f365" />

Reply:
Unfortunately this change introduced problems to the solution, such as deadlocks, some of them are already fixed, but I will have to revisit the entire PR in the future.